### ISSUE TITLE
[YCQL] flaky test CppCassandraDriverTest.ConcurrentIndexUpdate

### ISSUE THREAD

**Issue by m-iancu** [yugabyte/yugabyte-db#14556]
Jira Link: [DB-3939](https://yugabyte.atlassian.net/browse/DB-3939)
### Description

https://detective-gcp.dev.yugabyte.com/stability/test?branch=master&build_type=all&class=CppCassandraDriverTest&fail_tag=all&name=ConcurrentIndexUpdate&platform=all

Looks like most failures are a fail assert along the lines of:
```
../../src/yb/integration-tests/cassandra_cpp_driver-test.cc:2655
Value of: p.second
Actual: false
Expected: true
Duplicate key: 19, value: 1019, existing value: 14019
```

**Comment by rthallamko3** [comment]
From Oleg
```
Failed YCQL test:   CppCassandraDriverTest.ConcurrentIndexUpdate    (https://github.com/yugabyte/yugabyte-db/issues/14556)
Error:  Duplicate key: 18, value: 3018, existing value: 2018
Problem: the Index is NOT synced with the main table. Table change must remove a row from the Index and add another row. Due to a bug we have BOTH rows in the Index (the new one and the previous, which must be deleted).
The test (pseudo code):
CREATE TABLE test.key_value (key INT, value INT, PRIMARY KEY((key))) WITH transactions = {'enabled':true};
CREATE INDEX index_by_value ON test.key_value (value);

p = PREPARE( INSERT INTO test.key_value (key, value) VALUES (?, ?); )
for (int loop = 1; loop <= 20; ++loop)
    for (int key = 0; key != 10; ++key)
        EXECUTE( p.bind(key, loop*1000 + key) )     <<< parallel conflicting INSERT ops

r = select * from index_by_value;
ASSERT( r must have only one row for every 'key' )  <<< FAILED
After the test the table must have only 10 rows. The Index must have also only 10 rows accordingly.
But in fact the Index have > 10 rows due to the duplicates:
select * from index_by_value:
key: 6, value: 17006
key: 4, value: 4004
key: 1, value: 11001
key: 2, value: 5002
key: 0, value: 3000
key: 5, value: 17005
key: 1, value: 3001
!!! Duplicate key: 1, value: 3001, existing value: 11001
key: 8, value: 7008
key: 7, value: 20007
key: 9, value: 2009
key: 3, value: 8003
key: 6, value: 11006
!!! Duplicate key: 6, value: 11006, existing value: 17006
TOTAL: 12
The race is quite rare in fact: 1 / 10 for 200 parallel insertions minimum (600 is better to repro = 20 loops * 30 keys).
Which releases are affected?  Master, 2.16, 2.14 (< 05/20/2022), 2.12 (< 01/26/2022)
2.8, 2.6, 2.4, 2.2 do not have the error message (but there are At least one process crashed. viz: ts-N errors from the end of 2020).
The test was created on Feb 29, 2020.
Test details in detective:  [https://detective-gcp.dev.yugabyte.com/stability/test?branch=master&buckets=10&build_type=al[…]ndexUpdate&num_commits=500&num_fails=100&platform=all](https://detective-gcp.dev.yugabyte.com/stability/test?branch=master&buckets=10&build_type=all&class=CppCassandraDriverTest&fail_tag=all&name=ConcurrentIndexUpdate&num_commits=500&num_fails=100&platform=all)
Additional notes:
The Index update happens in DocDB (from QLWriteOperation::UpdateIndexes). CQL Proxy Executor only provides the Index ID to update it on the server side (it does not send a write-op directly to the Index tablet).
I can guess that the race is related to the transaction conflict resolution - in OptimisticLockingConflictResolver / ConflictResolver. May be to the AbortTransactions code path.
```

**Comment by robertsami** [comment]
```
diff --git a/src/yb/docdb/cql_operation.h b/src/yb/docdb/cql_operation.h
index ea68ad6176..95a6ebf298 100644
--- a/src/yb/docdb/cql_operation.h
+++ b/src/yb/docdb/cql_operation.h
@@ -13,6 +13,7 @@
 
 #pragma once
 
+#include <sstream>
 #include "yb/common/ql_protocol.pb.h"
 #include "yb/common/typedefs.h"
 
@@ -54,6 +55,41 @@ class QLWriteOperation :
     return index_requests_;
   }
 
+  std::string ToString() const override {
+    std::stringstream ss;
+    if (request().type() == QLWriteRequestPB_QLStmtType_QL_STMT_INSERT) {
+      ss << "INSERT";
+    } else if (request().type() == QLWriteRequestPB_QLStmtType_QL_STMT_DELETE) {
+      ss << "DELETE";
+    } else {
+      return "";
+    }
+    if (request().is_backfill()) {
+      ss << "-BACKFILL";
+    }
+    if (request().hashed_column_values_size() == 0 || (request().column_values_size() == 0 && request().range_column_values_size() == 0)) {
+      return "";
+    }
+    ss << "(";
+    ss << request().hashed_column_values(0).value().int32_value();
+    ss << ", ";
+    if (request().range_column_values_size() > 0) {
+      ss << request().range_column_values(0).value().int32_value();
+    } else {
+      ss << request().column_values(0).expr().value().int32_value();
+    }
+    ss << ")";
+    if (request().child_transaction_data().has_read_time()) {
+      ss << " @ " << request().child_transaction_data().read_time().ShortDebugString();
+    }
+    if (request().has_user_timestamp_usec()) {
+      HybridTime t(request().user_timestamp_usec());
+      ss << " // @ " << t;
+    }
+    // ss << " === " << request().ShortDebugString();
+    return ss.str();
+  }
+
   // Rowblock to return the "[applied]" status for conditional DML.
   const QLRowBlock* rowblock() const { return rowblock_.get(); }
 
diff --git a/src/yb/tablet/transaction_coordinator.cc b/src/yb/tablet/transaction_coordinator.cc
index a6ed96694c..eaae39bf96 100644
--- a/src/yb/tablet/transaction_coordinator.cc
+++ b/src/yb/tablet/transaction_coordinator.cc
@@ -768,6 +768,8 @@ class TransactionState {
     data.state.aborted().ToGoogleProtobuf(&aborted_);
     is_external_ = data.state.has_external_commit_ht();
 
+    LOG(INFO) << "COMMIT(" << id() << ") @ " << commit_time_;
+
     involved_tablets_.reserve(data.state.tablets().size());
     for (const auto& tablet : data.state.tablets()) {
       InvolvedTabletState state = {
diff --git a/src/yb/tablet/write_query.cc b/src/yb/tablet/write_query.cc
index 4bbbc40807..36f676a3cf 100644
--- a/src/yb/tablet/write_query.cc
+++ b/src/yb/tablet/write_query.cc
@@ -28,6 +28,7 @@
 #include "yb/docdb/conflict_resolution.h"
 #include "yb/docdb/consensus_frontier.h"
 #include "yb/docdb/cql_operation.h"
+#include "yb/docdb/docdb_debug.h"
 #include "yb/docdb/doc_write_batch.h"
 #include "yb/docdb/pgsql_operation.h"
 #include "yb/docdb/redis_operation.h"
@@ -213,6 +214,19 @@ void WriteQuery::Cancel(const Status& status) {
 }
 
 void WriteQuery::Complete(const Status& status) {
+  // std::stringstream ss;
+  // ss << "T " << (*tablet())->tablet_id() << " - ";
+  // auto& write_batch = request().write_batch();
+  // if (write_batch.has_transaction()) {
+  //   ss << CHECK_RESULT(FullyDecodeTransactionId(write_batch.transaction().transaction_id()));
+  //   ss << " -- ";
+  // }
+  // for (const auto& op : doc_ops_) {
+  //   ss << op->ToString() << " // ";
+  // }
+  // LOG(INFO) << ss.str() << " - COMPLETE";
+
+
   Release();
 
   if (callback_) {
@@ -417,6 +431,18 @@ void WriteQuery::Execute(std::unique_ptr<WriteQuery> query) {
 Status WriteQuery::DoExecute() {
   auto shared_tablet = VERIFY_RESULT(tablet());
   auto& write_batch = *request().mutable_write_batch();
+
+  std::stringstream ss;
+  ss << "T " << shared_tablet->tablet_id() << " - ";
+  if (write_batch.has_transaction()) {
+    ss << CHECK_RESULT(FullyDecodeTransactionId(write_batch.transaction().transaction_id()));
+    ss << " -- ";
+  }
+  for (const auto& op : doc_ops_) {
+    ss << op->ToString() << " // ";
+  }
+  LOG(INFO) << ss.str();
+
   isolation_level_ = VERIFY_RESULT(shared_tablet->GetIsolationLevelFromPB(write_batch));
   const RowMarkType row_mark_type = GetRowMarkTypeFromPB(write_batch);
   const auto& metadata = *shared_tablet->metadata();
diff --git a/src/yb/tserver/tablet_service.cc b/src/yb/tserver/tablet_service.cc
index 988214a1ed..4fcd129433 100644
--- a/src/yb/tserver/tablet_service.cc
+++ b/src/yb/tserver/tablet_service.cc
@@ -1726,6 +1726,14 @@ bool EmptyWriteBatch(const docdb::KeyValueWriteBatchPB& write_batch) {
 void TabletServiceImpl::Write(const WriteRequestPB* req,
                               WriteResponsePB* resp,
                               rpc::RpcContext context) {
+  for (const auto& op : req->ql_write_batch()) {
+    if (op.is_backfill()) {
+      HybridTime ht(req->read_time().read_ht());
+      LOG(INFO) << req->ShortDebugString() << " - " << ht;
+
+      break;
+    }
+  }
   if (FLAGS_TEST_tserver_noop_read_write) {
     for (int i = 0; i < req->ql_write_batch_size(); ++i) {
       resp->add_ql_response_batch();

```

with the diff above, saw the following:

```
...
[ts-3] I1118 15:04:10.004734 390987776 write_query.cc:444] T 6c4952c79f6f44cf8b4b561c87a16bf8 - 81ef49b3-9db1-4eda-8e94-89ab678f118e -- INSERT(2023, 23) //
[ts-2] I1118 15:04:10.305970 370130944 transaction_coordinator.cc:771] COMMIT(81ef49b3-9db1-4eda-8e94-89ab678f118e) @ { days: 19314 time: 15:04:10.049057 }
...
[ts-3] I1118 15:04:10.324903 413523968 write_query.cc:444] T 6c4952c79f6f44cf8b4b561c87a16bf8 - 4fab338a-f3dd-4a04-af1a-267a5f42bdc7 -- INSERT(4023, 23) // DELETE(2023, 23) //
[ts-2] I1118 15:04:10.782626 331497472 transaction_coordinator.cc:771] COMMIT(4fab338a-f3dd-4a04-af1a-267a5f42bdc7) @ { days: 19314 time: 15:04:10.390928 }
...
[ts-3] I1118 15:04:10.790764 451620864 tablet_service.cc:1732] tablet_id: "6c4952c79f6f44cf8b4b561c87a16bf8" include_trace: false ql_write_batch { type: QL_STMT_INSERT hash_code: 60333 hashed_column_values { value { int32_value: 1017 } } range_column_values { value { int32_value: 17 } } is_backfill: true } ql_write_batch { type: QL_STMT_INSERT hash_code: 47067 hashed_column_values { value { int32_value: 3019 } } range_column_values { value { int32_value: 19 } } is_backfill: true } ql_write_batch { type: QL_STMT_INSERT hash_code: 52104 hashed_column_values { value { int32_value: 2023 } } range_column_values { value { int32_value: 23 } } is_backfill: true } read_time { read_ht: 6835412380130615296 DEPRECATED_max_of_read_time_and_local_limit_ht: 6835412380130615296 global_limit_ht: 6835412380130615296 in_txn_limit_ht: 18446744073709551615 local_limit_ht: 6835412380130615296 } client_id1: 236060389406678413 client_id2: 14834195304433546626 request_id: 504 min_running_request_id: 501 rejection_score: 0 external_hybrid_time: 6835412380130615296 batch_idx: 18446744073709551615 - { days: 19314 time: 15:04:10.617826 }
[ts-3] I1118 15:04:10.791000 451620864 write_query.cc:444] T 6c4952c79f6f44cf8b4b561c87a16bf8 - INSERT-BACKFILL(1017, 17) // INSERT-BACKFILL(3019, 19) // INSERT-BACKFILL(2023, 23) //
...
```

tl;dr -- it looks like backfill is resurrecting data in the index
